text/0000-programmatic-style-injection.md

```diff
@@ -0,0 +1,172 @@
+import myStyles from './styles.css';
+
+export default Example class extends LightningElement {
+    static stylesheets = [
+        myStyles
+    ];
+}
+```
+
+## Motivation
+
+CSS custom properties and CSS `::part()` are the standard mechanisms by which applications can influence the look and feel of shadow DOM components. Component authors are required to explicitly define those extension points that can be styled from outside the shadow tree. While this approach gives a certain level of control over the component appearance, in certain cases developers might want to take full control over the styles injected in the shadow tree. And this without compromising the component encapsulation. 
+
+Component libraries are a prime example of this use case. Component authors want to expose high-quality base components with minimal styles and let application developers add their branding.
+
+LWC implicit template and stylesheet resolution system makes it impossible today to extend from a base LWC component and override its styles at the same time. This proposal addresses this issue by exposing a mechanism to associate stylesheets directly with the LWC component class. 
+
+## Detailed design
+
+A new `stylesheets` static property is added on the `LightningElement` constructor. This property allows developers to define which stylesheets the component should use. It accepts an array of [LWC stylesheets](#lwc-stylesheets-modules). The base `LightningElement.stylesheets` static property returns an empty array.
+
+The `stylesheets` property works on both light DOM and shadow DOM LWC components. This stylesheet injection mechanism complements the existing implicit template/stylesheet resolution mechanism.
+
+The LWC engine caches the static `stylesheets` value for the lifetime of the application during the component class definition. Modifying the `stylesheets` property after the component code is evaluated will not affect the stylesheet injected by this component. 
+
+For component inheritance, the LWC engine doesn't attempt to merge ancestor components stylesheets. Developers are in charge of merging (or not) the ancestors' stylesheets with their stylesheets.
+
+```js
+import { LightningElement } from 'lwc';
+import ancestorStylesheet from './ancestor.css';
+import childStylesheet from './child.css';
+
+class Ancestor extends LightningElement {
+    static stylesheets = [
+        ancestorStylesheet
+    ];
+}
+
+class Descendent extends Ancestor {
+    static stylesheets = [
+        ...super.stylesheets, 
+        childStylesheet
+    ];
+}
+```
+
+### Injection order
+
+The LWC engine injects stylesheets in the following order: first, the implicit stylesheet associated with the template and then all the stylesheets associated with the `stylesheets` property on the component class. This order allows the styles that are programmatically associated with the component to override the styles that are implicitly loaded via the template.

```

Comment:
In the case of multi-template components, we cannot guarantee that the implicit stylesheets associated with the template apply before the stylesheets from the `stylesheets` property on the component class. This is because we never remove templates and instead only append them.

So for instance, if you have `static stylesheets = [X]` and you `render()` template A with stylesheet A, then later `render` template B with stylesheet B, you would get this ordering:

1. `A`
2. `X`
3. `B`

To me, it is more consistent and predictable if instead the ordering were:

1. `X`
2. `A`
3. `B`

However, if we change this ordering, then it would no longer be true that you can "override the styles that are implicitly loaded via the template." So I am a bit torn on this.

Reply:
> In the case of multi-template components, we cannot guarantee that the implicit stylesheets associated with the template apply before the stylesheets from the `stylesheets` property on the component class. This is because we never remove templates and instead only append them.
> 
> So for instance, if you have `static stylesheets = [X]` and you `render()` template A with stylesheet A, then later `render()` template B with stylesheet B, you would get this ordering in the DOM:
> 
> 1. `A`
> 2. `X`
> 3. `B`
> 
> To me, it is more consistent and predictable if instead the ordering were:
> 
> 1. `X`
> 2. `A`
> 3. `B`
> 
> However, if we change this ordering, then it would no longer be true that you can "override the styles that are implicitly loaded via the template." So I am a bit torn on this.
> 
> As I note [here](https://github.com/salesforce/lwc-rfcs/pull/56#discussion_r981693420), there is also the issue that the order will never change on subsequent updates. (E.g. if we re-`render()` template A with stylesheet A, then none of the stylesheets in the DOM will change.)

If I'm thinking of the cascade, I think we want this order:

1. `A`
2. `B`
3. `X`

Provided the following example, i'd expect `X` stylesheet to to be apply last in the order:
```css
/* Template A CSS */
.a {
  color: var(--a-color);
}
```

```css
/* Template B CSS */
.b {
  color: var(--b-color);
}
```

```css
/* Template X CSS */
--a-color: red;
--b-color: blue;
```

Result is the `color` of `.a` is `red` and `.b` is `blue`.

Reply:
@brandonferrua In your example, the order doesn't matter, correct? [Codepen](https://codepen.io/nolanlawson-the-selector/pen/oNdEpKR)

![Screen Shot 2022-09-28 at 10 47 42 AM](https://user-images.githubusercontent.com/283842/192853034-311b6f1e-d456-4cac-871e-29a3956e81df.png)

No matter what order the `<style>`s are in, the result should be the same.

Reply:
It would if I have CSS property declarations, for example https://codepen.io/brandonferrua/pen/abGqqNM/512551f3612bdde4e0c228d1b5bbaf6d

Reply:
@brandonferrua Yes, that makes sense. I just wanted to make sure I understand what your use case is.

In order to get the order you want (`A B X`), we would need to insert stylesheets at a particular point in the DOM/`adoptedStyleSheets` rather than just blindly appending. (I.e. insert `B` between `A` and `X`.) Also as I mentioned, it doesn't solve the problem of "Rendering A, then B, then A again." (In that case, the order would still be `A B X`, with `B` after `A`.)

I am starting to wonder if the only way to truly solve this problem is to [remove unused stylesheets](https://github.com/salesforce/lwc/pull/2797). In that system, rendering B would remove A's stylesheets, and vice versa. (It was a small perf hit, but the PR proved it was doable.) Either that, or when running `render()`, we could change the order of stylesheets.

Reply:
Ah yes, I see the concern about "A, then B, then A again".

Removing the unused stylesheets seems like an elegant idea and a cleaner experience than dealing with `render()`. 

Could both implementations exist? Where the static property always is injected last and you remove unused stylesheets or someone chooses to be explicit, constructing their templates and stylesheets using the `render()` method for full control.

Reply:
> Could both implementations exist? Where the static property always is injected last and you remove unused stylesheets or someone chooses to be explicit, constructing their templates and stylesheets using the `render()` method for full control.

Sure, but in that case, would you still need the static property? You could always append whatever stylesheet you want to the end of the `stylesheets` you return from `render()`:

```js
render() {
  return {
    template,
    stylesheets: [otherStylesheet, stylesheetIWantLast]
  }
}
```

In my mind, the ordering for the above would be:

- `template`'s inherent stylesheets
- `otherStylesheet`
- `stylesheetIWantLast`

Reply:
If I'm looking at how much code the author would need to write to accomplish the same thing, I'd prefer the static property.

If I Subclass an ancestor Class:
```
import barStylesheets from 'bar.css';

class Bar extends Foo {
  static stylesheets = [...stylesheets, barStylesheets];
}
```

I don't need to import the HTML template plus the inherited stylesheet, I get that implicitly so that's nice.

If I could only use the render method, it gets a bit clunky:
```
import barTemplate from 'bar.html'
import fooStylesheets from '../foo/foo.css';
import barStylesheets from 'bar.css';

class Bar extends Foo {
  render() {
    return {
      template: barTemplate,
      stylesheets = [fooStylesheets, barStylesheets]
    }
  }
}
```
It's fine and it works but I feel like I'm having to "reset" things I otherwise should have gotten through inheritance.

Reply:
@brandonferrua Interesting. In my `render` proposal, though, I believe this should work?

```js
class Bar extends Foo {
  render() {
    const fooStylesheets = super.render().stylesheets
    return {
      template,
      stylesheets: [...fooStylesheets, ...barStylesheets]
  }
}
```

I agree though that it's less elegant than `static stylesheets = [...super.stylesheets, ...barStylesheets]`.

Reply:
Another wrinkle to add to this problem:

Even if we removed unused stylesheets, there is still a problem if we have multiple components of the same class on the same page.

For instance, imagine two multi-template components. The first may be in state A and the second in state B. If they are using a global styling system (synthetic shadow or light-DOM-scoped), then we can't have an ordering between the A and B stylesheets that satisfies the cascade needs for both components.

In terms of programmatic stylesheets, we would have the same issue if two components of the same class rendered different non-template stylesheets (i.e. ` render() { return { template, stylesheets } }`) in my proposal.

The only way to fully solve this may be to 1) have unique scope tokens per-instance of a component, and 2) remove unused stylesheets.

Reply:
Another thought: this problem already exists today with multi-template components. Perhaps we need a separate RFC to address it.

And even without addressing that use case: leveraging the `render()` function to return the `stylesheets` should (at least) replace the [current hacks](https://github.com/salesforce/lwc-test/pull/147), and it would still work for multi-template components with no conflicts between the different stylesheets.